### PR TITLE
Otters - Ruge Li

### PR DESCRIPTION
**Pull request by rugeli**

### CODE REVIEW

File: tests/test_wave_03.py [hunk 1]
```diff
@@ -54,12 +54,13 @@ def test_friends_unique_movies_not_duplicated():
 
     # Arrange
     assert len(friends_unique_movies) == 3
+    assert INTRIGUE_3 in friends_unique_movies
```
Comment (goeunpark) [inline tests/test_wave_03.py:57]: I would like to see more assertions here to truly test that `friend_unique_movies` do not have duplicates, as it describes to do so on L47. It's customary for our test definition to describe in full detail what our test block intends to test. Currently, this assertion only says that `INTRIGUE_3` is in `friends_unique_movies`, it doesn't actually test that `friends_unique_movies` has three distinct movies!

File: tests/test_wave_05.py [hunk 2]
```diff
@@ -57,7 +57,7 @@ def test_new_genre_rec_from_empty_friends():
     # ****** Complete the Act and Assert Portions of theis tests **********
```
Comment (goeunpark) [inline tests/test_wave_05.py:57]: Make sure to check the test file thoroughly so you don't miss additional instructions! (The `README.md` for these projects are unfortunately very long but has important info!)

File: viewing_party/party.py [hunk 3]
```diff
@@ -1,23 +1,143 @@
...
 def create_movie(title, genre, rating):
-    pass
+    movie_dict = {
+        "title": title,
+        "genre": genre,
+        "rating": rating
+    }
+    for value in movie_dict.values():
+        if value == None:
+            return None
```
Comment (goeunpark) [inline viewing_party/party.py:11]: Great validation check! 

Additionally, It's an even better idea to run checks like this before we take up memory with a possibly unnecessary variable -- we could move this checkto L4, above your `movie_dict` dictionary, and return None if title or genre or movie == None.

File: viewing_party/party.py [hunk 4]
```diff
@@ -1,23 +1,143 @@
...
+            user_data["watched"].append(movie)
+            user_data["watchlist"].remove(movie)
+    return user_data      
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    sum_rating = 0.0
+    if len(user_data["watched"]) == 0:
+        avg_rating = 0.0
```
Comment (goeunpark) [inline viewing_party/party.py:35]: A few things here: 

- `avg_rating` is placed inside this conditional code block and the program accesses this variable _outside_ of this if block in L39. This works in Python because the smallest local scope for variables are within a function. However, this will not work in many other languages (including Javascript, which we'll learn in Unit 3!) because the scope of this `avg_rating` would be within this `if` block. (If this is confusing -- don't worry, this stumped me for a good 20 minutes today!) 
- Given all that, it's usually good practice to define variables at its broadest "layer"! I would personally pull `avg_rating = 0.0` above L34 to avoid this possible confusion!
- We set `avg_rating` as zero if the watched list is empty -- here we can just `return 0`!

File: viewing_party/party.py [hunk 5]
```diff
@@ -1,23 +1,143 @@
...
+        sum_rating += movie["rating"]
+        avg_rating = sum_rating/len(user_data["watched"])
+    return avg_rating
 
-
+def get_most_watched_genre(user_data):
+    popular_genre = ""
+    genres = []
+    if len(user_data["watched"]) == 0:
+        popular_genre = None
```
Comment (goeunpark) [inline viewing_party/party.py:45]: Nice validation check from the beginning! ✨  Similar to the comment above, we can even `return None` here to pop out of this function if the watched list is zero!

File: viewing_party/party.py [hunk 6]
```diff
@@ -1,23 +1,143 @@
...
+    for movie in user_data["watched"]:
+        genres.append(movie["genre"])
+        popular_genre = max(set(genres), key=genres.count)
```
Comment (goeunpark) [inline viewing_party/party.py:48]: A really great combo of `max()` and `set()` on this line!

File: viewing_party/party.py [hunk 7]
```diff
@@ -1,23 +1,143 @@
...
+    return popular_genre 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
-
+def get_unique_watched(user_data):
+    unique_watched = []
+    friend_watched = []
+    user_watched = []
```
Comment (goeunpark) [inline viewing_party/party.py:56]: Setting up all of our local variables at the top of the function is a great coding decision because it makes it so much easier for the reader!

File: viewing_party/party.py [hunk 8]
```diff
@@ -1,23 +1,143 @@
...
+    if len(user_data["watched"]) == 0:
+        return unique_watched
+    for watched_movie in user_data["watched"]:
+        user_watched.append(watched_movie)
+    for friend_watched_dict in user_data["friends"]:
+        for friend_watched_movie in friend_watched_dict["watched"]:
+            friend_watched.append(friend_watched_movie)
+    for movie in user_watched:
+        if movie not in friend_watched:
+            unique_watched.append(movie)
```
Comment (goeunpark) [inline viewing_party/party.py:66]: This is some really dense for-loop logic -- I think you did it elegantly and in the right order! Two things that might make this easier to read is to make comments and a line break in front of the three main for-loops, like so:

```
    # get user's movie titles
   for watched_movie in user_data["watched"]:
        user_watched.append(watched_movie)

    # get friends' movie titles
    for friend_watched_dict in user_data["friends"]:
        for friend_watched_movie in friend_watched_dict["watched"]:
            friend_watched.append(friend_watched_movie)

    # check for unique titles
    for movie in user_watched:
        if movie not in friend_watched:
            unique_watched.append(movie) 
```

File: viewing_party/party.py [hunk 9]
```diff
@@ -1,23 +1,143 @@
...
+    user_watched = []
+    movies_set = set()
+    for watched_movie in user_data["watched"]:
+        user_watched.append(watched_movie)
+    for friend_watched_dict in user_data["friends"]:
+        for friend_watched_movie in friend_watched_dict["watched"]:
+            friend_watched.append(friend_watched_movie)
+    for movie in friend_watched:
+        if movie not in user_watched:
+            friends_unique_movies.append(movie)
```
Comment (goeunpark) [inline viewing_party/party.py:81]: 👀  Oooh, I see an opportunity to refactor and DRY this up a little bit here -- this block is pretty much the same as L57-L66! Maybe there could be a way to make a lil' helper function! 🤔

File: viewing_party/party.py [hunk 10]
```diff
@@ -1,23 +1,143 @@
...
+    final_friends_unique_movies = []
+    for movie in friends_unique_movies:
+        each_movie = tuple(movie.items())
```
Comment (goeunpark) [inline viewing_party/party.py:84]: Sprinkling a tuple in here, I love it!

File: viewing_party/party.py [hunk 11]
```diff
@@ -1,23 +1,143 @@
...
+            final_friends_unique_movies.append(movie)
+    return final_friends_unique_movies
+
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    rec_movies = []
+    friend_watched = get_friends_unique_watched(user_data)
```
Comment (goeunpark) [inline viewing_party/party.py:96]: Excellent use of using a helper method! 🌟

File: viewing_party/party.py [hunk 12]
```diff
@@ -1,23 +1,143 @@
...
+
+
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    rec_movies_by_genre = []
+    most_watched_genre = get_most_watched_genre(user_data)
+    friend_watched = get_friends_unique_watched(user_data)
```
Comment (goeunpark) [inline viewing_party/party.py:118]: Great use of helper functions! 🎉

File: viewing_party/party.py [hunk 13]
```diff
@@ -1,23 +1,143 @@
...
+    user_watched = []
+    total_movies = []
```
Comment (goeunpark) [inline viewing_party/party.py:120]: I noticed that for this function and for the `get_available_recs()` function above, you combine the user's and friends' movies by creating a a `total_movies` list! This totally works, but there are ways to reduce time and space complexity and compare the movies lists without creating a new combined list. Instead of L121-L129, consider checking for the genre and appending to `rec_movies_by_genre` list in one for-loop:

```
for movie in friend_watched:
    if movie['genre'] == most_watched_genre:
        rec_movies_by_genre.append(movie)
```

**Review by goeunpark — COMMENTED** [review]
Excellent work, Ruge! 🎉 

The tests pass, your code jumps through for-loops without dropping a sweat, and your use of helper functions are wonderful! I can see that you're comfortable keying into a dict and creating new lists to create data structures that you want. You missed a testing requirement on Wave 05 but I'm thinking that was an accident? Let me know if you want to practice or go through writing Act and Assert statements. Overall, this project is a Green! 🟢 

I've left a few comments below that could make your code clearer and more efficient. Let me know if you have any questions about this feedback or any of the explanations (especially in regards to how scoping in Python is different from other languages)!

Great job on your first project!
